src/longsequences/copying.jl

```diff
@@ -7,28 +7,109 @@
+"""
+    copy!(dst::LongSequence, src::BioSequence)
+
+In-place copy content of `src` to `dst`, resizing `dst` to fit.
+The alphabets of `src` and `dst` must be compatible.
+
+# Examples
+```
+julia> seq = copy!(dna"TAG", dna"AACGTM")
+6nt DNA Sequence:
+AACGTM
+
+julia> copy!(seq, rna"UUAG")
+4nt DNA Sequence:
+TTAG
+```
+"""
+function Base.copy!(dst::LongSequence{A}, src::LongSequence{A}) where {A <: Alphabet}
+    return _copy!(dst, src)
+end
+
+function Base.copy!(dst::LongSequence{<:NucleicAcidAlphabet{N}},
+                    src::LongSequence{<:NucleicAcidAlphabet{N}}) where N
+    return _copy!(dst, src)
+end
+
+function _copy!(dst::LongSequence, src::LongSequence)
+    orphan!(dst)
+    # Most efficient
+    if !src.shared
+        resize!(dst.data, length(src.data))
+        copyto!(dst.data, src.data)
+        dst.part = src.part
+    # Less efficient
+    else
+        resize!(dst, length(src))
+        @inbounds copyto!(dst, src)
+    end
+    return dst
+end
+
+"""
+    copyto!(dst::LongSequence, src::BioSequence)
+
+Equivalent to `copyto!(dst, 1, src, 1, length(src))`
+"""
+function Base.copyto!(dst::LongSequence{A}, src::LongSequence{A}) where {A <: Alphabet}
+    return copyto!(dst, 1, src, 1, length(src))
+end
+
+function Base.copyto!(dst::LongSequence{<:NucleicAcidAlphabet{N}},
+                      src::LongSequence{<:NucleicAcidAlphabet{N}}) where N
+    return copyto!(dst, 1, src, 1, length(src))
 end
 
+"""
+    copyto!(dst::LongSequence, soff, src::BioSequence, doff, N)
+
+In-place copy `N` elements from `src` starting at `soff` to `dst`, starting at `doff`.
+The length of `dst` must be greater than or equal to `N + doff - 1`.
+The first N elements of `dst` are overwritten,
+the other elements are left untouched. The alphabets of `src` and `dst` must be compatible.
+
+# Examples
+```
+julia> seq = copyto!(dna"AACGTM", 1, dna"TAG", 1, 3)
+6nt DNA Sequence:
+TAGGTM
+
+julia> copyto!(seq, 2, rna"UUUU", 1, 4)
+6nt DNA Sequence:
+TTTTTM
+```
+"""
 function Base.copyto!(dst::LongSequence{A}, doff::Integer,
-                    src::LongSequence{A}, soff::Integer) where {A}
-    return copyto!(dst, doff, src, soff, length(src) - soff + 1)
+                      src::LongSequence{A}, soff::Integer,
+                      N::Integer) where {A <: Alphabet}
+    return _copyto!(dst, doff, src, soff, N)
+end
+
+function Base.copyto!(dst::LongSequence{<:NucleicAcidAlphabet{B}}, doff::Integer,
+                      src::LongSequence{<:NucleicAcidAlphabet{B}}, soff::Integer,
+                      N::Integer) where B
+    return _copyto!(dst, doff, src, soff, N)
 end
 
-@inline function Base.copyto!(dst::LongSequence{A}, doff::Integer,
-                    src::LongSequence{A}, soff::Integer, len::Integer) where {A}
-    @boundscheck checkbounds(dst, doff:doff+len-1)
-    @boundscheck checkbounds(src, soff:soff+len-1)
+function _copyto!(dst::LongSequence{A}, doff::Integer,
+                  src::LongSequence, soff::Integer,
```

Comment:
This is for simplicity, since we may allow a different source and destination alphabet, e.g. `DNAAlphabet{2}` and `RNAAlphabet{2}`. `A` is only used to determine the number of bits per nucleotide.